wasm/js/skia_renderer.js

```diff
@@ -4,47 +4,54 @@ Module.onRuntimeInitialized = function () {
         this._ctx = canvas.getContext('2d');
 
         this._realRenderer = _offscreenGL;
-        this['clear'] = function () {
-            const canvas = this._canvas;
-            const glCanvas = _offscreenGL._canvas;
 
-            if (glCanvas.width < canvas.width || glCanvas.height < canvas.height) {
-                glCanvas.width = canvas.width;
-                glCanvas.height = canvas.height;
+        this._drawList = [];
+        this._saveCount = 0;
+
+        this['clear'] = function () {
+            this._drawList = [];  // TODO: what happens on clear if there is a clip???
+            _pendingRenderers.delete(this);
+            if (this._saveCount != 0) {
+                throw "Dunno how to handle a save count in clear()";
             }
-            _offscreenGL.clear();
         };
 
         this['save'] = function () {
-            _offscreenGL.save();
+            ++this._saveCount;
+            this._drawList.push(_offscreenGL['save'].bind(_offscreenGL));
         };
 
         this['restore'] = function () {
-            _offscreenGL.restore();
+            if (this._saveCount > 0) {
+                --this._saveCount;
+                this._drawList.push(_offscreenGL['restore'].bind(_offscreenGL));
+            }
         };
 
         this['transform'] = function (xform) {
-            _offscreenGL.transform(xform);
+            this._drawList.push(_offscreenGL['transform'].bind(_offscreenGL, xform));
         };
 
         this['align'] = function (fit, align, from, to) {
-            _offscreenGL.align(fit, align, from, to);
+            this._drawList.push(_offscreenGL['align'].bind(_offscreenGL, fit, align, from, to));
         };
 
         this['computeAlignment'] = function (mat2d, fit, align, from, to) {
-            _offscreenGL.computeAlignment(mat2d, fit, align, from, to);
+            this._drawList.push(_offscreenGL['computeAlignment'].bind(
+                    _offscreenGL, mat2d, fit, align, from, to));
         };
 
         this['flush'] = function () {
-            _offscreenGL.flush();
-            const ctx = this._ctx;
-            ctx.globalCompositeOperation = 'copy';
-            ctx.drawImage(_offscreenGL._canvas, 0, 0);
+            while (this._saveCount > 0) {
+                this['restore']();
+            }
+            if (_pendingRenderers.size == 0) {
+                setTimeout(flushPendingRenderers, 0);
```

Comment:
I really dislike setting a timeout here. Could we alternatively (now or later) require the user to call "rive.flushOffscreenCanvas()' during their animation callback, once they're done operating on all their offscreen canvases?

Reply:
Done.

After cleaning up the requestAnimationFrame stuff, this isn't as good as I had hoped, but still a 2x boost on safari.